agent/cache-types/streaming_health_services_test.go

```diff
@@ -0,0 +1,492 @@
+			gatherNodes(result.Value))
+
+		opts.MinIndex = result.Index
+		opts.LastResult = &result
+	})
+
+	runStep(t, "batched updates work too", func(t *testing.T) {
+		// Simulate multiple registrations happening in one Txn (so all have same
+		// index)
+		batchEv := newEventBatchWithEvents(
+			// Deregister an existing node
+			newEventServiceHealthDeregister(20, 1, "web"),
+			// Register another
+			newEventServiceHealthRegister(20, 4, "web"),
+		)
+		client.QueueEvents(batchEv)
+		opts.Timeout = time.Second
+		result, err := typ.Fetch(opts, req)
+		require.NoError(t, err)
+
+		require.Equal(t, uint64(20), result.Index)
+		require.ElementsMatch(t, []string{"node2", "node3", "node4"},
+			gatherNodes(result.Value))
+
+		opts.MinIndex = result.Index
+		opts.LastResult = &result
+	})
+}
+
+func TestStreamingHealthServices_Filtering(t *testing.T) {
+	client := NewTestStreamingClient()
+	typ := StreamingHealthServices{deps: MaterializerDeps{
+		Client: client,
+		Logger: hclog.Default(),
+	}}
+
+	// Create an initial snapshot of 3 instances but in a single event batch
+	batchEv := newEventBatchWithEvents(
+		newEventServiceHealthRegister(5, 1, "web"),
+		newEventServiceHealthRegister(5, 2, "web"),
+		newEventServiceHealthRegister(5, 3, "web"))
+	client.QueueEvents(
+		batchEv,
+		newEndOfSnapshotEvent(pbsubscribe.Topic_ServiceHealth, 5))
+
+	// This contains the view state so important we share it between calls.
+	opts := cache.FetchOptions{
+		MinIndex: 0,
+		Timeout:  1 * time.Second,
+	}
+	req := &structs.ServiceSpecificRequest{
+		Datacenter:  "dc1",
+		ServiceName: "web",
+		QueryOptions: structs.QueryOptions{
+			Filter: `Node.Node == "node2"`,
+		},
+	}
+
+	gatherNodes := func(res interface{}) []string {
+		nodes := make([]string, 0, 3)
+		r := res.(*structs.IndexedCheckServiceNodes)
+		for _, csn := range r.Nodes {
+			nodes = append(nodes, csn.Node.Node)
+		}
+		return nodes
+	}
+
+	runStep(t, "filtered snapshot returned", func(t *testing.T) {
+		result, err := typ.Fetch(opts, req)
+		require.NoError(t, err)
+
+		require.Equal(t, uint64(5), result.Index)
+		require.Equal(t, []string{"node2"}, gatherNodes(result.Value))
+
+		opts.MinIndex = result.Index
+		opts.LastResult = &result
+	})
+
+	runStep(t, "filtered updates work too", func(t *testing.T) {
+		// Simulate multiple registrations happening in one Txn (so all have same
+		// index)
+		batchEv := newEventBatchWithEvents(
+			// Deregister an existing node
+			newEventServiceHealthDeregister(20, 1, "web"),
+			// Register another
+			newEventServiceHealthRegister(20, 4, "web"),
+		)
+		client.QueueEvents(batchEv)
+		opts.Timeout = time.Second
+		result, err := typ.Fetch(opts, req)
+		require.NoError(t, err)
+
+		require.Equal(t, uint64(20), result.Index)
+		require.Equal(t, []string{"node2"}, gatherNodes(result.Value))
+
+		opts.MinIndex = result.Index
+		opts.LastResult = &result
+	})
+}
+
+func runStep(t *testing.T, name string, fn func(t *testing.T)) {
+	if !t.Run(name, fn) {
```

Comment:
Technically correct, although since this function does not call `t.Log`, `t.Error`, or `t.Fatal`, and does not pass `t` to any other functions, I think it would be a no-op to call `t.Helper` here. But It doesn't hurt either, so I'll add it in case that changes.